File: cs101/pa1/List.java
// List.java

public class List {

   // private inner Node class
   @SuppressWarnings("overrides")
   private class Node {
      // Fields
      int data;
      Node next;
      Node prev;

      // constructor
      Node(int data) {
         this.data = data;
         next = prev = null;
      }

      // toString():  overrides Object's toString() method
      // copy from Queue.java example
      public String toString() {
         return String.valueOf(data);
      }

      //equals(): overrides Object's equals() method
      //copy from Queue.java example
      public boolean equals(Object x) {
         boolean eq = false;
         Node that;
         if(x instanceof Node) {
            that = (Node) x;
            eq = (this.data == that.data);
         }
         return eq;
      }
   }

   // Fields
   private Node front;
   private Node back;
   private Node present;
   private int length;
   private int cursorIndex;

   // Constructor
   List() { // Creates a new empty list.
      front = null;
      back = null;
      present = null;
      cursorIndex = -1;
      length = 0;
   }

   // access functions
   
   // Returns the number of elements in this List
   int length() {
      return length;
   }

   // If cursor is defined, returns the index of the cursor element,
   // otherwise returns -1;
   int index() {
      if(length() > 0 ) {
         return cursorIndex;
      }else {
         return -1; // means cursor is undefined
      }
   }

   // Returns front element. Pre: length()>0
   int front() {
      if(length() > 0) {
         return front.data;
      }else {
         throw new RuntimeException("List Error: front() called on empty List");
      }
   }

   // Returns back element. Pre: length()>0
   int back() {
      if(length() <= 0) {
         throw new RuntimeException("List Error: back() called on empty List");
      }
      return back.data;
   }

   // Returns cursor element. Pre: length()>0, index()>=0
   int get() {
      if(length() <= 0) {
         throw new RuntimeException("List Error: get() called on empty List");
      }else if(present == null){
         throw new RuntimeException("List Error: get() called on undefined cursor");
      }else {
         return present.data;
      }
   }

   // Returns true if and only if this List and L are the same
   // integer sequence. This states of the cursors in the two Lists 
   // are not used in determinging equality
   // copy from Queue.java example and modified
   boolean equals(List L) {
      boolean eq = false;
      List Q;
      Node N, M;
      if(L instanceof List) {
         Q = L;
         N = this.front;
         M = Q.front;
         eq = (this.length == Q.length);
         while(eq && N != null) {
            eq = N.equals(M);
            N = N.next;
            M = M.next;
         }
      }
      return eq;
   }

   // Manipulation procedures

   // Resets this List to its original empty state.
   void clear() {
      front = null;
      back = null;
      present = null;
      length = 0;
      cursorIndex = -1;
   }

   // If List is non-empty, places the cursor under the front element,
   // otherwise does nothing.
   void moveFront() {
      if(length() > 0) {
         present = front;
         cursorIndex = 0; // index starts from 0;
      }

   }

   // If List is non-empty, places the cursor under the back element,
   // otherwise does nothing.
   void moveBack() {
      if(length() > 0) {
         present = back;
         cursorIndex = length -1;
      }
   }

   // If cursor is defined and not at front, moves cursor one step toward
   // front of this List, if cursor is defiend and at front, cursor becomes
   // undefined, if cursor is undefined does nothing.
   void movePrev() {
      if(present == front) { // if cursor is defined and at front
         present = null;
         cursorIndex = -1; // make cursor undefine
      }else { 
         present = present.prev;
         cursorIndex--;
      }
   }



   // If cursor is defined and not at back, moves cursor one step toward 
   // back of this List, if cursor is defiend and at back, cursor becomes
   // undefined, if cursor is undefined does nothing.
   void moveNext() {
      if(present == back) {
         present = null;
         cursorIndex = -1;
      }else {
         present = present.next;
         cursorIndex++;
      }
   }

   // Insert new element into this List. If List is non-empty, 
   // insertion takes place before front element.
   void prepend(int data) {
      Node N = new Node(data);
      if(length() <= 0) {
         front = N;
         back = N;
      }else {
         front.prev = N;
         N.next = front;
         N.prev = null;
         front = N;
      }
      cursorIndex++;
      length++;
   }

   // Insert new element into this List. If List is non-empty.
   // insertion takes place after back element.
   // copy from Queue.java example
   void append(int data) {
      Node N = new Node(data);
      if(length() <= 0) {
         front = N;
         back = N;
      }else {
         back.next = N;
         N.prev = back; // when it reaches the end, it should points back
         back = N;
      }
      length++;
   }

   // Insert new element before cursor.
   // Pre: length()>0, index()>=0
   void insertBefore(int data) {
      if(length() > 0 && index() >= 0) {
         if(present == front) {
            prepend(data);
         }else {
            Node N = new Node(data);
            N.prev = present.prev;
            N.next = present;
            present.prev.next = N;
            present.prev = N;
            length++;
            cursorIndex++;
         }
      }else {
         throw new RuntimeException("List Error: insertBefore() called on empty List");
      }
   }

   // Insert new element after cursor.
   // Pre: length()>0, index()>0
   void insertAfter(int data) {
      if(length() > 0 && index() >= 0) {
         if(present == back) {
            append(data);
         }else {
            Node N = new Node(data);
            N.prev = present;
            N.next = present.next;
            N.next.prev = N;
            present.next = N;
            length++;
         }
      }else {
         throw new RuntimeException("List Error: insertafter() called on empty List");
      }
   }

   // Deletes the front element. Pre: length()>0
   void deleteFront() {
      if(length() <= 0){
         throw new RuntimeException("List Error: deleteFront() called on empty List");
      }else if(back == front) {
         back = front = null;
         present = null;
         cursorIndex = -1;
      }else{
         front = front.next;
         front.prev = null;
         cursorIndex--;
      }
      //cursorIndex = -1;
      length--;
   }

   // Delete the back element. Pre: length()>0
   void deleteBack() {
      if(length() <= 0) {
         throw new RuntimeException("List Error: deleteBack() called on empty List");
      }
      if(front == back) {
         front = back = null;
         present = null;
         cursorIndex = -1; 
      }
      if(present == back){
         present = null;
         cursorIndex = -1;
      }else{
         back = back.prev;
         back.next = null;
      }
      length--;
   }

   // Deletes cursor element, making cursor undefined
   // Pre: length()>0, index()>=0
   void delete() {
      if(length() <= 0 && index() < 0) {
         throw new RuntimeException("List Error: delete() called on empty List");
      }
      if(present == back){
         deleteBack();
         back = back.prev;
      }else if(present == front) {
         deleteFront();
      }else {
         present.prev.next = present.next;
         present.next.prev = present.prev;
         present = null;
         cursorIndex = -1;
         length--;
      }
   }

   //other methods

   // Overrides Object's toString method. Returns a String
   // representation of this List consisting of a space
   // separated sequence of integers, with front on left.
   // copy from Queue.java and modified
   public String toString() {
      StringBuffer sb = new StringBuffer();
      Node N = front;
      while( N != null) {
         sb.append(N.toString());
         sb.append(" ");
         N = N.next;
      }
      return new String(sb);
   }

   // Returns a new List representing the same integer sequence as this 
   // List. The cursor in the new List is undefined, regardless of the 
   // state of the cursor in this List and L. THe states of this List and L 
   // are unchanged
   // copy from Queue.java example
    List copy() {
      List L = new List();
      Node N = this.front;

      while(N != null) {
         L.append(N.data);
         N = N.next;
      }
      return L;
   }
}

File: cs101/pa2/List.c
// List.c
// Implementation file for List ADT

#include<stdio.h>
#include<stdlib.h>
#include"List.h"

// structs

// private NodeObj type
typedef struct NodeObj {
   int data;
   struct NodeObj* next;
   struct NodeObj* prev;
} NodeObj;

// private Node type
typedef NodeObj* Node;

Node newNode(int data) {
   Node N = malloc(sizeof(NodeObj));
   N->data = data;
   N->next = NULL;
   N->prev = NULL;
   return(N);
}

void freeNode(Node* pN) {
   if(pN != NULL && pN != NULL) {
      free(*pN);
      *pN = NULL;
   }
}
// private ListObj type
typedef struct ListObj {
   Node front;
   Node back;
   Node curr;
   int length;
   int cursorIndex;
} ListObj;

// Constructor-Destrucotrs

// newList()
// Returns reference to new empty List object.
List newList(void) {
   List L = malloc(sizeof(ListObj));
   L->front = L->back = L->curr = NULL;
   L->cursorIndex = -1;
   L->length = 0;
   return (L);
}


// freeList()
// Frees all heap memory associated with List *pL, and sets *pL to NUll.
// alter from Queue.c example
void freeList(List* pL) {
   if(pL != NULL && *pL != NULL) {
      Node N = (*pL)->front;
      while(N != NULL) {
         Node curr = N;
         N = N->next;
         free(curr);
      }
      free(*pL);
      *pL = NULL;
   }
}

// Accrss functions

// length()
// Returns the length of L.
int length(List L) {
   if(L == NULL) {
      printf("List Error: calling length() on NULL List reference.\n");
      exit(EXIT_FAILURE);
   }
   return L->length;
}

// index()
// Returns the index of cursor.
// If curosr is defined, returns the index of the cursor element,
// otherwise returns -1;
int index(List L) {
   if(L == NULL) {
      printf("List Error: calling index() calling on NULL List reference.\n");
      exit(EXIT_FAILURE);
   }
   if(L->curr == NULL) {
      return -1;
   }

   if(length(L) > 0) {
      return L->cursorIndex;
   }else {
      return -1;
   }
}

// front()
// Returns the front element of L.
// Pre: length() > 0
int front(List L) {
   if(L == NULL) {
      printf("List Error: calling front() on NULL List reference.\n");
      exit(EXIT_FAILURE);
   }
   if(length(L) <= 0) {
      printf("List Error: calling front() on an empty List.\n");
      exit(EXIT_FAILURE);
   }
   return L->front->data;
}

// back()
// Returns the back element of L.
// Pre: length()>0
int back(List L) {
   if(L == NULL) {
      printf("List Error: calling back() on NULL List reference.\n");
      exit(EXIT_FAILURE);
   }
   if(length(L) <= 0) {
      printf("List Error: calling back() on an empty List.\n");
      exit(EXIT_FAILURE);
   }
   return L->back->data;
}

// get()
// Returns the present elemnt that cursor locates.
// Pre: length()>0, index()>=0
int get(List L) {
   if(L == NULL) {
      printf("List Error: calling get() on NULL List.\n");
      exit(EXIT_FAILURE);
   }
   if(length(L) <= 0) {
      printf("List Error: calling get() on an empty List.\n");
      exit(EXIT_FAILURE);
   }
   if(index(L) < 0) {
      printf("List Error: calling get() on an NULL cursor List.\n");
      exit(EXIT_FAILURE);
   }
   return L->curr->data;
}

// equals()
// Returns true (1) if A is identical to B, false (0) otherwise
// copy from Queue.c example
int equals(List A, List B) {
   int eq = 0;
   Node N = NULL;
   Node M = NULL;
   if( A == NULL || B == NULL ){
      printf("List Error: calling equals() on NULL List reference.\n");
      exit(EXIT_FAILURE);
   }
   eq = ( A->length == B->length );
   N = A->front;
   M = B->front;
   while( eq && N!=NULL){
      eq = (N->data == M->data);
      N = N->next;
      M = M->next;
   }
   return eq;
}

// Manipulation procedures

// clear()
// Make L become empty.
void clear(List L) {
   if(L == NULL) {
      printf("List Error: calling clear() on NULL List reference.\n");
      exit(EXIT_FAILURE);
   }
   Node N = L->front;
   while(N != NULL) {
      Node temp = N;
      N = N->next;
      freeNode(&temp);
   }
   L->front = NULL;
   L->back = NULL;
   L->curr = NULL;
   L->length = 0;
   L->cursorIndex = -1;
}

// moveFront()
// Move the cursor locate at the first element
void moveFront(List L) {
   if(L == NULL) {
      printf("List Error: calling moveBack() on NULL List reference.\n");
      exit(EXIT_FAILURE);
   }
   if(length(L) <= 0) {
      printf("List Error: calling moveBack() on an empty List.\n");
      exit(EXIT_FAILURE);
   }else {
      L->curr = L->front;
      L->cursorIndex = 0;
   }
}

// moveBack()
// Move the cursor locate at the last element
void moveBack(List L) {
   if(L == NULL) {
      printf("List Error: calling moveBack() on NULL List reference.\n");
      exit(EXIT_FAILURE);
   }
   if(length(L) <= 0) {
      printf("List Error: calling moveBack() on an empty List.\n");
      exit(EXIT_FAILURE);
   }else {
      L->curr = L->back;
      L->cursorIndex = (L->length) - 1;
   }
}

// movePrev()
// Make the cursor locate at the element in front of it.
void movePrev(List L){
   if(L == NULL) {
      printf("List Error: calling movePrev() on NULL List reference.\n");
      exit(EXIT_FAILURE);
   }
   if(length(L) <= 0) {
      printf("List Error: calling movePrev() on an empty List.\n");
      exit(EXIT_FAILURE);
   }
   if(L->curr == L->front) {
      L->curr = NULL;
      freeNode(&L->curr);
      L->cursorIndex = -1;
   }else {
      L->curr = L->curr->prev;
      L->cursorIndex--;
   }
}

// moveNext()
// Make the cursor locate at the element after it.
void moveNext(List L) {
   if(L == NULL) {
      printf("List Error: calling moveNext() on NULL List reference.\n");
      exit(EXIT_FAILURE);
   }
   if(length(L) <= 0) {
      printf("List Error: calling moveNext() on an empty List.\n");
      exit(EXIT_FAILURE);
   }
   if(L->curr == L->back) {
      L->curr = NULL;
      freeNode(&L->curr);
      L->cursorIndex = -1;
   }else {
      L->curr = L->curr->next;
      L->cursorIndex++;
   }
}

// prepend()
// Insert an elment at the head of the List.
void prepend(List L, int data) {
   if(L == NULL) {
      printf("List Error: calling prepend() on NULL List reference.\n");
      exit(EXIT_FAILURE);
   }
   Node N = newNode(data);
   if(length(L) == 0) {
      L->front = L->back = N;
      N->prev = N->next = NULL;
   }else {
      L->front->prev = N;
      N->next = L->front;
      L->front = N;
      N->prev = NULL;
   }
   if(L->curr != NULL) {
      L->cursorIndex++;
   }
   L->length++;
}

// append()
// Insert an element at the end of the List.
void append(List L, int data) {
   if(L == NULL) {
      printf("List Error: calling append() on NUll List reference.\n");
      exit(EXIT_FAILURE);
   }
   Node N = newNode(data);
   if(length(L) <= 0) {
      L->front = L->back = N;
      N->next = N->prev = NULL;
   }else {
      L->back->next = N;
      N->prev = L->back;
      L->back = N;
      L->back->next = NULL;
   }
   L->length++;
}

// insertBefore()
// Insert an element in front of the cursor element.
void insertBefore(List L, int data) {
   if(L == NULL) {
      printf("List Error: calling insertBefore() on NULL List reference.\n");
      exit(EXIT_FAILURE);
   }
   if(length(L) <= 0) {
      printf("List Error: calling insertBefore() on an empty List.\n");
      exit(EXIT_FAILURE);
   }
   if(index(L) < 0) {
      printf("List Error: calling insertBefore() on undefined cursor.\n");
      exit(EXIT_FAILURE);
   }
   if(L->curr == L->front) {
      prepend(L, data);
   }else {
      Node N = newNode(data);
      N->prev = L->curr->prev;
      N->next = L->curr;
      L->curr->prev->next = N;
      L->curr->prev = N;
      L->length++;
      L->cursorIndex++;
   }
}

// insertAfter
// Insert an element after the cursor element.
void insertAfter(List L, int data) {
   if(L == NULL) {
      printf("List Error: calling insertAfter() on NULL List reference.\n");
      exit(EXIT_FAILURE);
   }
   if(length(L) <= 0) {
      printf("List Error: calling insertAfter() on an empty List.\n");
      exit(EXIT_FAILURE);
   }
   if(index(L) < 0) {
      printf("List Error: calling insertAfter() on undefined cursor.\n");
      exit(EXIT_FAILURE);
   }
   if(L->curr == L->back) {
      append(L, data);
   }else {
      Node N = newNode(data);
      N->prev = L->curr;
      N->next = L->curr->next;
      N->next->prev = N;
      L->curr->next = N;
      L->length++;
   }
}

// deleteFront()
// Delete the front element
void deleteFront(List L) {
   if( L == NULL) {
      printf("List Error: calling deleteFront() on NULL List reference.\n");
      exit(EXIT_FAILURE);
   }
   if(length(L) == 0) {
      printf("List Error: calling deleteFront() on an empty List.\n");
      exit(EXIT_FAILURE);
   }
   Node N = NULL;
   N = L->front;
   if(length(L) > 1) {
      L->front = L->front->next;
      L->front->prev = NULL;
   }else {
      L->front = L->back = NULL;
   }
   L->length--;
   if(L->curr != NULL) {
      L->cursorIndex--;
   } else {
      L->cursorIndex = -1;
      //free(L->curr);
   }
   freeNode(&N);
}

// deleteBack()
// Delete the back element
void deleteBack(List L) {
   if(L == NULL) {
      printf("List Error: calling deleteback() on NULL List reference.\n");
      exit(EXIT_FAILURE);
   }
   if(length(L) <= 0) {
      printf("List Error: calling deleteBack() on an empty List.\n");
      exit(EXIT_FAILURE);
   }
   Node N = NULL;
   N = L->back;
   if(L->curr == L->back) {
      L->curr = NULL;
      L->cursorIndex = -1;
      //free(L->curr);
   }
   if(length(L) > 1) {
      L->back = L->back->prev;
      L->back->next = NULL;
   }else {
      L->front = L->back = NULL;
   }
   L->length--;
   freeNode(&N);
}

// delete()
// Delete the current element that cursor locates
void delete(List L) {
   if(L == NULL) {
      printf("List Error: calling delete() on NULL List reference.\n");
      exit(EXIT_FAILURE);
   }
      if(length(L) <= 0) {
      printf("List Error: calling delete() on an empty List.\n");
      exit(EXIT_FAILURE);
   }
   if(index(L) < 0) {
      printf("List Error: calling delete() on undefined cursor.\n");
      exit(EXIT_FAILURE);
   }
   if(L->curr == L->back) {
      deleteBack(L);
   }else if(L->curr == L->front) {
      deleteFront(L);
   }else {
      L->curr->prev->next = L->curr->next;
      L->curr->next->prev = L->curr->prev;
      L->curr = NULL;
      L->cursorIndex = -1;
      L->length--;
      freeNode(&L->curr);

   }
}

// Other operations

// PrintList()
// Prints data elements in L on a single line to stdout.
void printList(FILE* out, List L) {
   Node N = NULL;
   if(L==NULL){
      printf("List Error: calling printList() on NULL List reference.\n");
      exit(EXIT_FAILURE);
   }

   for(N = L->front; N != NULL; N = N->next){
      fprintf(out,"%d ", N->data);
   }
}

// copyList()
// Create a new List by copy from L.
List copyList(List L) {
   List A = newList();
   Node N = L->front;
   while(N != NULL) {
      append(A, N->data);
      N = N->next;
   }
   return A;
}
